File: src/main/java/net/atos/zac/admin/model/ZaaktypeCmmnZaakafzenderParameters.java
/*
 * SPDX-FileCopyrightText: 2023 Atos, 2025 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.zac.admin.model;

import static nl.info.zac.database.flyway.FlywayIntegrator.SCHEMA;

import java.util.Objects;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.SequenceGenerator;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;

@Entity
@Table(schema = SCHEMA, name = "zaaktype_cmmn_zaakafzender_parameters")
@SequenceGenerator(schema = SCHEMA, name = "sq_zaaktype_cmmn_zaakafzender_parameters", sequenceName = "sq_zaaktype_cmmn_zaakafzender_parameters", allocationSize = 1)
public class ZaaktypeCmmnZaakafzenderParameters implements UserModifiable<ZaaktypeCmmnZaakafzenderParameters> {

    public enum SpecialMail {
        GEMEENTE,
        MEDEWERKER;

        public boolean is(final String name) {
            return this.name().equals(name);
        }
    }

    @Id
    @GeneratedValue(generator = "sq_zaaktype_cmmn_zaakafzender_parameters", strategy = GenerationType.SEQUENCE)
    @Column(name = "id")
    private Long id;

    @NotNull @ManyToOne
    @JoinColumn(name = "zaaktype_configuration_id", referencedColumnName = "id")
    private ZaaktypeCmmnConfiguration zaaktypeCmmnConfiguration;

    @Column(name = "default_mail", nullable = false)
    private boolean defaultMail;

    @NotBlank @Column(name = "mail", nullable = false)
    private String mail;

    @Column(name = "replyto")
    private String replyTo;

    public Long getId() {
        return id;
    }

    public void setId(final Long id) {
        this.id = id;
    }

    public ZaaktypeCmmnConfiguration getZaaktypeCmmnConfiguration() {
        return zaaktypeCmmnConfiguration;
    }

    public void setZaaktypeCmmnConfiguration(final ZaaktypeCmmnConfiguration zaaktypeCmmnConfiguration) {
        this.zaaktypeCmmnConfiguration = zaaktypeCmmnConfiguration;
    }

    public boolean isDefault() {
        return defaultMail;
    }

    public void setDefault(final boolean defaultMail) {
        this.defaultMail = defaultMail;
    }

    public String getMail() {
        return mail;
    }

    public void setMail(final String mail) {
        this.mail = mail;
    }

    public String getReplyTo() {
        return replyTo;
    }

    public void setReplyTo(final String replyTo) {
        this.replyTo = replyTo;
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof ZaaktypeCmmnZaakafzenderParameters that))
            return false;
        return defaultMail == that.defaultMail && Objects.equals(mail, that.mail) && Objects.equals(replyTo, that.replyTo);
    }

    @Override
    public int hashCode() {
        return Objects.hash(defaultMail, mail, replyTo);
    }

    @Override
    public boolean isModifiedFrom(ZaaktypeCmmnZaakafzenderParameters original) {
        return Objects.equals(mail, original.mail) && (!defaultMail == original.defaultMail ||
                                                       !Objects.equals(replyTo, original.replyTo));
    }

    @Override
    public void applyChanges(ZaaktypeCmmnZaakafzenderParameters changes) {
        this.defaultMail = changes.defaultMail;
        this.replyTo = changes.replyTo;
    }

    @Override
    public ZaaktypeCmmnZaakafzenderParameters resetId() {
        id = null;
        return this;
    }
}

File: src/main/kotlin/nl/info/zac/admin/model/ZaaktypeCmmnZaakafzenderParameters.kt
/*
 * SPDX-FileCopyrightText: 2025 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.zac.admin.model

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.SequenceGenerator
import jakarta.persistence.Table
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import nl.info.zac.database.flyway.FlywayIntegrator
import nl.info.zac.util.AllOpen
import java.util.Objects

@Entity
@Table(schema = FlywayIntegrator.SCHEMA, name = "zaaktype_cmmn_zaakafzender_parameters")
@SequenceGenerator(
    schema = FlywayIntegrator.SCHEMA,
    name = "sq_zaaktype_cmmn_zaakafzender_parameters",
    sequenceName = "sq_zaaktype_cmmn_zaakafzender_parameters",
    allocationSize = 1
)
@AllOpen
class ZaaktypeCmmnZaakafzenderParameters : UserModifiable<ZaaktypeCmmnZaakafzenderParameters> {

    enum class SpecialMail {
        GEMEENTE,
        MEDEWERKER;

        fun name(name: String): Boolean = this.name == name
    }

    @Id
    @GeneratedValue(generator = "sq_zaaktype_cmmn_zaakafzender_parameters", strategy = GenerationType.SEQUENCE)
    @Column(name = "id")
    var id: Long? = null

    @ManyToOne
    @JoinColumn(name = "zaaktype_configuration_id", referencedColumnName = "id")
    @field:NotNull
    lateinit var zaaktypeCmmnConfiguration: ZaaktypeCmmnConfiguration

    @Column(name = "default_mail", nullable = false)
    var defaultMail: Boolean = false

    @Column(name = "mail", nullable = false)
    @field:NotBlank
    lateinit var mail: String

    @Column(name = "replyto")
    @field:NotBlank
    lateinit var replyTo: String

    override fun isModifiedFrom(original: ZaaktypeCmmnZaakafzenderParameters): Boolean {
        return Objects.equals(mail, original.mail) && (
            !defaultMail == original.defaultMail ||
                !Objects.equals(replyTo, original.replyTo)
            )
    }

    override fun applyChanges(changes: ZaaktypeCmmnZaakafzenderParameters) {
        this.defaultMail = changes.defaultMail
        this.replyTo = changes.replyTo
    }

    override fun resetId(): ZaaktypeCmmnZaakafzenderParameters {
        id = null
        return this
    }

    override fun equals(other: Any?): Boolean {
        if (other !is ZaaktypeCmmnZaakafzenderParameters) return false
        return mail == other.mail && defaultMail == other.defaultMail && Objects.equals(replyTo, other.replyTo)
    }

    override fun hashCode(): Int {
        return Objects.hash(mail, defaultMail, replyTo)
    }
}
